types/validator_set.go

```diff
@@ -366,6 +335,204 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
 	}
 }
 
+// Checks changes against duplicates, splits the changes in updates and removals, sorts them by address
+func processChanges(origChanges []*Validator) (updates, removals []*Validator, err error) {
```

Comment:
Maybe changes can be split into additions, removals and updates. The potential benefit is the fact that we can easily determine size of the new valset, and we can probably also get rid of `Remove` function (like we get rid of `Add`) function.

Reply:
I added an applyRemove2() function to show how it would look without Remove().

Reply:
Is `split into additions, removals and updates` related to `also get rid of Remove function` ? 

Also you guys inspired me to write the shorted applyRemove I could. Check it out below!

Reply:
> split into additions, removals and updates

This sounds good, but it might make things more complicated in `updateWithChangeSet` since we have to do a few things for both additions and updates. Not sure what's best.

Reply:
We probably don't need to separate updates further, but I have other proposal. What if instead of `applyUpdates` and `applyRemovals` we have single function that just applies updates and computes nextValidatorSet. This function would start from empty val set (size is equal to existing.size + additions.size - deletions.size) and then iterate over existing valset and for each validator check if it is update, addition or deletion, and then do appropriate logic. For deletions we don't do anything; for updates and additions, we add it to the new set and compute proposer priority (in place; I am not sure `computeNewPriorites` function brings a lot). To make this possible, updates and deletions should be maps, and `processChanges` would need to return size of the new val set. Verify methods would just return error and not compute modified voting power as they do in the last version, so it might be a bit cleaner. `Remove` method is not needed anymore.